app/controllers/profile_controller.ts

```diff
@@ -0,0 +1,215 @@
+import type { HttpContext } from '@adonisjs/core/http'
+import {
+  loadUserWithTwoFactor,
+  userHasWebauthnCredentials,
+  markSecurityConfirmed,
+  isSecurityConfirmed,
+  SECURITY_CONFIRMATION_CHALLENGE_KEY,
+  parseTransports,
+  parseRecoveryCodes,
+  generateAndStoreTwoFactorSecret,
+} from '#services/two_factor'
+import { getRpId, getOrigin, fromBase64Url } from '#services/webauthn_service'
+import { confirmSecurityValidator } from '#validators/profile_validator'
+import { db } from '#services/db'
+import hash from '@adonisjs/core/services/hash'
+import { webauthnServer } from '#services/webauthn_server'
+import type { AuthenticationResponseJSON } from '@simplewebauthn/types'
+
+export default class ProfileController {
+  async show({ auth, inertia }: HttpContext) {
+    const user = await loadUserWithTwoFactor(auth.user!.id)
+    const hasWebauthn = await userHasWebauthnCredentials(user.id)
+    const recoveryCodes = parseRecoveryCodes(user.twoFactorRecoveryCodes)
```

Comment:
Do we really need a method for parsing the recovery codes? Don't we have full control over the format? - e.g. we can know for sure that it's an array of strings because that's what we inserted outselves into the database.